ecc/bls12-377/fr/sis/sis.go

```diff
@@ -107,362 +93,245 @@ func NewRSis(seed int64, logTwoDegree, logTwoBound, maxNbElementsToHash int) (*R
+func (r *RSis) InnerHash(it *LimbIterator, res, k, kz fr.Vector, polId int, mask uint64) {
+	copy(k, kz)
+	zero := uint64(0)
+
+	// perf note: there is room here for additional improvement with the mask.
+	// for example, since we already know some of the "rows" of field elements are going to be zero
+	// we could have an iterator that "skips" theses rows and avoid func call / buffer fillings.
+	// also, we could update the mask if some non-const rows happens to be zeroes,
+	// such that the FFT we select has less work to do (in some cases; i.e. we need a bunch
+	// of following limbs to be zero to make it worth it).
+
+	for j := 0; j < r.Degree; j++ {
+		l, ok := it.NextLimb()
+		if !ok {
+			break
 		}
-		return h
-	}, nil
-
+		zero |= l
+		k[j][0] = l
+	}
+	if zero == 0 {
+		// means m[i*r.Degree : (i+1)*r.Degree] == [0...0]
+		// we can skip this, FFT(0) = 0
+		return
+	}
+	// this is equivalent to:
+	// 	r.Domain.FFT(k, fft.DIF, fft.OnCoset(), fft.WithNbTasks(1))
+	r.smallFFT(k, mask)
+
+	// we compute k * r.Ag[polId] in ℤ_{p}[X]/Xᵈ+1.
+	// k and r.Ag[polId] are in evaluation form on √(g) * <g>
+	// we accumulate the result in res; the FFT inverse is done once every multiplications are done.
+	k.Mul(k, fr.Vector(r.Ag[polId]))
+	res.Add(res, k)
 }
 
-func genRandom(seed, i, j int64, buf *bytes.Buffer) fr.Element {
-
-	buf.Reset()
-	buf.WriteString("SIS")
-	binary.Write(buf, binary.BigEndian, seed)
-	binary.Write(buf, binary.BigEndian, i)
-	binary.Write(buf, binary.BigEndian, j)
+func deriveRandomElementFromSeed(seed, i, j int64) fr.Element {
+	var buf [3 + 3*8]byte
+	copy(buf[:3], "SIS")
+	binary.BigEndian.PutUint64(buf[3:], uint64(seed))
+	binary.BigEndian.PutUint64(buf[11:], uint64(i))
+	binary.BigEndian.PutUint64(buf[19:], uint64(j))
 
-	digest := blake2b.Sum256(buf.Bytes())
+	digest := blake2b.Sum256(buf[:])
 
 	var res fr.Element
 	res.SetBytes(digest[:])
 
 	return res
 }
 
-// mulMod computes p * q in ℤ_{p}[X]/Xᵈ+1.
-// Is assumed that pLagrangeShifted and qLagrangeShifted are of the correct sizes
-// and that they are in evaluation form on √(g) * <g>
-// The result is not FFTinversed. The fft inverse is done once every
-// multiplications are done.
-func mulMod(pLagrangeCosetBitReversed, qLagrangeCosetBitReversed []fr.Element) []fr.Element {
-
-	res := make([]fr.Element, len(pLagrangeCosetBitReversed))
-	for i := 0; i < len(pLagrangeCosetBitReversed); i++ {
-		res[i].Mul(&pLagrangeCosetBitReversed[i], &qLagrangeCosetBitReversed[i])
-	}
-
-	// NOT fft inv for now, wait until every part of the keys have been multiplied
-	// r.Domain.FFTInverse(res, fft.DIT, true)
-
-	return res
+// TODO @gbotrel explore generic perf impact + go 1.23 iterators
+// i.e. the limb iterator could use generics and be instantiated with uint8, uint16, uint32, uint64
+// the iterators could implement the go 1.23 iterator pattern.
 
+// ElementIterator is an iterator over a stream of field elements.
+type ElementIterator interface {
```

Comment:
What is the purpose of this interface ? Are you anticipating other iterators than just the moving to the next fr.Element ?

Reply:
Yes, typically in `linea` see [this PR](https://github.com/Consensys/linea-monorepo/pull/554) we iterate over a matrix represented with `smartvectors`